src/VideoStreaming/VideoReceiver.cc

```diff
@@ -966,6 +968,61 @@ VideoReceiver::_keyframeWatch(GstPad* pad, GstPadProbeInfo* info, gpointer user_
 }
 #endif
 
+void VideoReceiver::takeSnapshot() {
+#if defined(QGC_GST_STREAMING)
+    qDebug() << "Taking Snapshot";
+    if (!_videoSink) {
+        return;
+    }
+
+    qDebug() << "Trying to take a snapshot";
+    GstSample *videobuffer = nullptr;
+    GstCaps *caps;
+    gint width, height;
+    GstMapInfo map;
+
+    g_object_get(G_OBJECT(_videoSink), "last-sample", &videobuffer, NULL);
+    if (!videobuffer) {
+        qDebug() << "No video buffer, exiting";
```

Comment:
This file contains a debug category, maybe you should use it.

Reply:
It will not work like that in case of qmlglsink packaged into GstBin (see https://github.com/mavlink/qgroundcontrol/blob/master/src/FlightDisplay/VideoManager.cc#L287-L379). To make that working you need to add some property to GstBin (let's call is also "last-sample") and forward call to qmlglsink inside that bin.

Reply:
> This file contains a debug category, maybe you should use it.

I'm planing to cleanup logging in whole VideoReceiver - as for me @tcanabrava can leave that as is, I'll take care of that.

Reply:
And one more thing - you need to analyze GstVideoInfo (you can get it from sink caps and sample's pool caps) to get raw pointer to the video buffer. Sometimes pixel 0, 0 is not aligned with begging of memory block (sample buffer as try to do) and depending on hardware and GL implementation it is not necessary that image row length will be equal to sizeof(pixel) * width - sometimes it is longer because of buffer aligning reasons. You need to use thing called pitch/stride to find beginning of next pixel row.

IMO the best possible implementation in your case (where performance does not really matter but maintenance and stability is the key) - it is to find some gstreamer tool that converts GstSample into some image evaluating GstVideoInfo, as you need. It is not big fun to deal with implementation-depended buffer aligning. I suspect that in your case it will be 99.9% aligned as you expect but can't guarantee it for all cases.